### PR TITLE
fix(controller): make managed namespaces more 'prune-proof'

### PR DESCRIPTION
**Pull request by blakepettersson**
In the cases where someone would want to set resource tracking on a managed namespace, or if someone would want to migrate from having a namespace in Git to using `managedNamespaceMetadata`, we need to take steps to ensure that the namespace does not get pruned.

In order to do that, we add the live namespace to the list of `targetObjs` (so that it's considered a part of the source even though it's not).

Finally, we need to also ensure that the managed namespace is not considered `OutOfSync` (due to the same reason as above).

This is a subset of #11350, the main difference being that this commit does not set resource tracking on its own, but can be opted into if people choose to do so.

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: controller/state.go [hunk 1]
```diff
@@ -597,12 +627,21 @@ func (m *appStateManager) CompareAppState(app *v1alpha1.Application, project *v1
 		} else {
 			diffResult = diff.DiffResult{Modified: false, NormalizedLive: []byte("{}"), PredictedLive: []byte("{}")}
 		}
+
+		// For the case when a namespace is managed with `managedNamespaceMetadata` AND it has resource tracking
+		// enabled (e.g. someone manually adds resource tracking labels or annotations), we need to do some
+		// bookkeeping in order to ensure that it's not considered `OutOfSync` (since it does not exist in source
+		// control).
+		//
+		// This is in addition to the bookkeeping we do above (see line ~500) to prevent said namespace from being pruned.
```
Comment (jsoref) [inline controller/state.go:636]: I'd argue against using `line` numbers for references.

You could try to use a code-ish path, e.g.:

> See the above `argo.GetTrackingMethod` loop that checks `app.Spec.SyncPolicy.ManagedNamespaceMetadata != nil`

Comment (blakepettersson) [reply controller/state.go:636]: Yeah that's a good point, I didn't really like this approach myself but couldn't think of a better way. I'll amend the comment 👍

**Review by ishitasequeira — APPROVED** [review]

**Comment by blakepettersson** [comment]
Thanks @jannfis! 🤗

**Pull request merged**